src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/zk/ZooKeeperSpout.java

```diff
@@ -0,0 +1,105 @@
+/* Copyright 2020 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+
+package org.openkilda.wfm.share.zk;
+
+import org.openkilda.bluegreen.LifeCycleObserver;
+import org.openkilda.bluegreen.LifecycleEvent;
+import org.openkilda.bluegreen.Signal;
+import org.openkilda.bluegreen.ZkWatchDog;
+import org.openkilda.wfm.AbstractBolt;
+import org.openkilda.wfm.CommandContext;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.storm.spout.SpoutOutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.topology.OutputFieldsDeclarer;
+import org.apache.storm.topology.base.BaseRichSpout;
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Values;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.UUID;
+
+@Slf4j
+public class ZooKeeperSpout extends BaseRichSpout implements LifeCycleObserver {
+    public static final String BOLT_ID = "zookeeper.spout";
+    public static final String FIELD_ID_LIFECYCLE_EVENT = "lifecycle.event";
+
+    public static final String FIELD_ID_CONTEXT = AbstractBolt.FIELD_ID_CONTEXT;
+    private String id;
+    private String serviceName;
+    private String connectionString;
+
+    private ZkWatchDog watchDog;
+    private SpoutOutputCollector collector;
+    private LifecycleEvent event;
+    private boolean newEvent = false;
+    private long messageId = 0;
+
+
+    public ZooKeeperSpout(String id, String serviceName, String connectionString) {
+        this.id = id;
+        this.serviceName = serviceName;
+        this.connectionString = connectionString;
+    }
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        this.collector = collector;
+        try {
+            this.watchDog = ZkWatchDog.builder().id(id).serviceName(serviceName)
+                    .connectionString(connectionString).build();
+            watchDog.subscribe(this);
+        } catch (IOException e) {
+            log.error("Failed to init ZooKeeper with connection string: {} received: {} ", connectionString,
+                    e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void nextTuple() {
+        if (event != null && newEvent) {
```

Comment:
What it received 2 events during waiting?

1. ZK send `event1`
2. we store it into `event` field
3. ZK send `event2`
4. we store it into `event` field
5. Someone ask spout to pull message.
6. we send only `event2` because we replaced `event1` on step 4

I think we need synchronized queue here
```